File: java-design-pattern/src/main/java/intro/to/design/patterns/using/javalang/iterator/IteratorMain.java
package intro.to.design.patterns.using.javalang.iterator;

import java.util.ArrayList;

public class IteratorMain {
    public static void main(String[] args) {
        BookShelf bookShelf = new BookShelf(4);
        bookShelf.appendBook(new Book("Around the World in 80 days"));
        bookShelf.appendBook(new Book("Bible"));
        bookShelf.appendBook(new Book("Cinderella"));
        bookShelf.appendBook(new Book("Daddy-Long-Legs"));

        Iterator<Book> it = bookShelf.iterator();
        while (it.hasNext()) {
            Book book = it.next();
            System.out.println(book.getName());
        }
        System.out.println();

        ArrayList<Book> books = bookShelf.getBooks();
        for (Book book : books) {
            System.out.println(book.getName());
        }
        System.out.println();
    }
}

File: c++-design-pattern/src/intro_to_design_patterns_using_java_lang/iterator/IteratorMain.cpp
#include "IteratorMain.h"

#include <iostream>
#include "Book.h"
#include "BookShelf.h"

IteratorMain::IteratorMain() = default;

IteratorMain::~IteratorMain() = default;

void IteratorMain::run()
{
    int maxSize{4};
    auto bookShelf = BookShelf{maxSize};
    bookShelf.appendBook(Book{"Around the World in 80 days"});
    bookShelf.appendBook(Book{"Bible"});
    bookShelf.appendBook(Book{"Cinderella"});
    bookShelf.appendBook(Book{"Daddy-Long-Legs"});

    auto it = bookShelf.iterator();
    while (it->hasNext())
    {
        Book book = it->next();
        std::cout << book.getName() << "\n";
    }
    std::cout << "\n";

    for (const auto &books = *(bookShelf.getBooks()); const auto &book : books)
    {
        std::cout << book.getName() << "\n";
    }
    std::cout << "\n";
}
